### PR TITLE
iOS merge DnD implementation stubs

### PR DESCRIPTION
**Pull request by elijah-semyonov**
Cherry-pick iOS DnD stubs and move `CMPGestureRecognizer` internals into Kotlin world.

### CODE REVIEW

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPDragInteractionProxy.m [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * limitations under the License.
+ */
+
+#import "CMPDragInteractionProxy.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation CMPDragInteractionProxy
+
++ (NSItemProvider *)itemProviderFromString:(NSString *)string {
```
Comment (ASalavei) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPDragInteractionProxy.m]: Can't find usages of the method.

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMacros.h [hunk 2]
```diff
@@ -16,14 +16,22 @@
...
-#define CMP_MUST_BE_OVERRIDED_INVARIANT_VIOLATION assert(false && "MUST_OVERRIDE");
+/// Marker for indicating that function is abstract and is expected to be overrided by subclasses
+#define CMP_ABSTRACT_FUNCTION
 
-// Marker for indicating that raw pointer returned from a function is owned by the caller
+/// Macro for assertion failure due to function marked as `CMP_ABSTRACT_FUNCTION`  wasn't overrided or called `super` implementation
+#define CMP_ABSTRACT_FUNCTION_CALLED assert(false && "This function must be overrided, and not call super implementation");
+
+/// Marker for functions which can be override, but have default behavior
+ #define CMP_CAN_OVERRIDE
```
Comment (ASalavei) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMacros.h:26]: extra space here

Comment (ASalavei) [inline compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPMacros.h]:
```suggestion
#define CMP_CAN_OVERRIDE
```

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        val parameters = UIPreviewParameters()
+        val cornerRadius = decorationCgRect.useContents {
+            minOf(size.width / 2.0, size.height  / 2.0, 4.0)
+        }
+        val path = UIBezierPath.bezierPathWithRoundedRect(decorationCgRect.useContents {
+            CGRectMake(0.0, 0.0, size.width, size.height)
+        }, cornerRadius = cornerRadius)
+        parameters.backgroundColor = UIColor.clearColor
+        parameters.visiblePath = path
+        parameters.visiblePath = path
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: duplicated line

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                dragAndDropTarget.onDrop(event)
+            }
+        }
+
+        override fun proposalForSessionUpdate(
+            session: UIDropSessionProtocol, interaction: UIDropInteraction
+        ): UIDropProposal = withDropSessionContext {
+            dragAndDropTarget.onMoved(event)
+            if (dragAndDropTarget.hasEligibleDropTarget) {
+                println("Has target")
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: excessed println

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                UIDropProposal(UIDropOperationCopy)
+            } else {
+                println("Has no target")
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: Same here

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import platform.UIKit.UIColor
+import platform.CoreGraphics.CGImageRelease
+import platform.CoreGraphics.kCGImageByteOrder32Little
+import platform.UIKit.UIImage
+
+/**
+ * Convert [androidx.compose.ui.graphics.Color] to iOS UIKit [UIColor]
+ * Assumes that source color is in sRGB color space.
+ */
+fun Color.toUIColor() = UIColor.colorWithRed(
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt]: internal?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    )?.releasedAfterScopeEnds() ?: return@withCFReleaseScope null
+
+    val cgImage = CGBitmapContextCreateImage(context) // must be released by the user
+    return@withCFReleaseScope cgImage
+}
+
+/**
+ * Creates a iOS UIKit [UIImage] with the contents of [ImageBitmap].
+ */
+fun ImageBitmap.toUIImage(): UIImage? = withCFReleaseScope {
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt]: Internal?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt [hunk 8]
```diff
@@ -117,6 +124,58 @@ private sealed interface UserInputViewHitTestResult {
...
+        // Can't check if either view is null
+        val view = gestureRecognizer.view ?: return false
+        val otherView = otherGestureRecognizer.view ?: return false
+
+        val otherIsAscendant = !otherView.isDescendantOfView(view)
+
+        // Only allow simultaneous recognition if the other gesture recognizer is attached to the same view
+        // or to a view up in the hierarchy
+        val result = otherView == view || otherIsAscendant
+        //            println("${gestureRecognizer.className} recognizes with ${otherGestureRecognizer.className}: $result")
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt]: debug code?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt [hunk 9]
```diff
@@ -117,6 +124,58 @@ private sealed interface UserInputViewHitTestResult {
...
+    ): Boolean {
+        // Other gesture recognizers,
+        // except the case where it belongs to the same view,
+        // are required to wait until we fail.
+        // In practice, it can only happen when other gesture recognizers are attached to the
+        // descendant views (aka interop views).
+        // In other cases, it's allowed to recognize simultaneously, so this method will not be
+        // called
+        val result = gestureRecognizer.view != otherGestureRecognizer.view
+        return result
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt]:
```suggestion
        return gestureRecognizer.view != otherGestureRecognizer.view
```

**Review by ASalavei — COMMENTED** [review]
@elijah-semyonov , did you have intention to fix swipe back in current MR as well? If yes, then unfortunately it does not work. I believe this is the kind of things that should work by default.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt (hunk 4 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: Fixed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt (hunk 5 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: Fixed

File: compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPDragInteractionProxy.m (hunk 1 above)
Comment (elijah-semyonov) [reply compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPDragInteractionProxy.m]: Aha, dead code indeed.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt (hunk 3 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/draganddrop/UIKitDragAndDropManager.uikit.kt]: Fixed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt (hunk 6 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt]: Fixed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt (hunk 7 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/Conversions.uikit.kt]: Fixed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt (hunk 8 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/UserInputView.uikit.kt]: Fixed

**Comment by elijah-semyonov** [comment]
> @elijah-semyonov , did you have intention to fix swipe back in current MR as well? If yes, then unfortunately it does not work. I believe this is the kind of things that should work by default.

@ASalavei you mean interactive navigation pop?

**Comment by ASalavei** [comment]
> @ASalavei you mean interactive navigation pop?

@elijah-semyonov , yes

**Review by ASalavei — APPROVED** [review]

**Comment by elijah-semyonov** [comment]
No, I didn't have this intention

**Pull request merged**